File: frontend/src/types/domain.ts
// Domain types aligned with backend paper_image/domain.py

export type CorpusScope = "public" | "private" | "submission";

export type EvidenceLevel =
  | "exact_duplicate"
  | "near_duplicate"
  | "partial_reuse"
  | "possible_manipulation"
  | "needs_review";

export type ReviewStatus =
  | "pending"
  | "confirmed"
  | "false_positive"
  | "needs_more_review"
  | "expert_review_requested";

export interface User {
  id: string;
  email: string;
  name: string;
  organization: string | null;
  role: "author" | "reviewer" | "admin";
  createdAt: string;
}

export interface Project {
  id: string;
  name: string;
  description: string;
  status: "draft" | "uploading" | "processing" | "ready" | "completed";
  targetPdf: PdfFile | null;
  referencePdfs: PdfFile[];
  createdAt: string;
  updatedAt: string;
}

export interface PdfFile {
  id: string;
  filename: string;
  fileSize: number;
  role: "target" | "reference";
  source: "upload" | "pubmed";
  uploadProgress: number;
  processingStatus: "pending" | "extracting" | "completed" | "failed";
  figureCount: number | null;
  panelCount: number | null;
  uploadedAt: string;
}

export interface Figure {
  id: string;
  articleId: string;
  page: number;
  caption: string;
  imagePath: string;
  bbox: [number, number, number, number];
  panels: Panel[];
}

export interface Panel {
  id: string;
  figureId: string;
  label: string;
  imagePath: string;
  bbox: [number, number, number, number];
  confirmed: boolean;
  confidence: number;
}

export interface PanelSummary {
  targetPanelCount: number;
  referencePanelCount: number;
  confirmedCount: number;
  pendingCount: number;
}

export interface Finding {
  id: string;
  level: EvidenceLevel;
  score: number;
  algorithm: string;
  message: string;
  targetPanel: PanelRef;
  sourcePanel: PanelRef | null;
  heatmapPath: string | null;
  overlayPath: string | null;
  modelVersion: string;
  parameters: Record<string, unknown>;
  reviewStatus: ReviewStatus;
  reviewedBy: string | null;
  reviewedAt: string | null;
}

export interface PanelRef {
  panelId: string;
  figureId: string;
  articleId: string;
  imagePath: string;
  label: string;
  scope: CorpusScope;
}

export interface ScreeningReport {
  id: string;
  projectId: string;
  generatedAt: string;
  summary: ReportSummary;
  findings: Finding[];
  panelUsage: PanelUsage;
  modelVersions: ModelVersion[];
}

export interface ReportSummary {
  totalFindings: number;
  byLevel: Record<EvidenceLevel, number>;
  highestRiskLevel: EvidenceLevel;
  figuresAnalyzed: number;
  panelsAnalyzed: number;
}

export interface PanelUsage {
  targetPanels: number;
  internalReferencePanels: number;
  externalReferencePanels: number;
  pubmedReferencePanels: number;
  estimatedCost: number | null;
}

export interface ModelVersion {
  module: string;
  version: string;
  name: string;
}

export interface Annotation {
  id: string;
  findingId: string;
  status: ReviewStatus;
  note: string;
  reviewerName: string;
  createdAt: string;
}

export interface ChatMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  citations: Citation[];
  createdAt: string;
  tokenCount?: number;
}

export interface Citation {
  findingId: string;
  figureId: string;
  text: string;
}

export interface PubMedSearchResult {
  pmid: string;
  title: string;
  authors: string[];
  journal: string;
  year: number;
  abstract: string;
  doi: string | null;
  pmcId: string | null;
  pdfAvailable: boolean;
}

File: src/paper_image/domain.py
from __future__ import annotations

import hashlib
from dataclasses import dataclass
from enum import Enum


class CorpusScope(str, Enum):
    PUBLIC = "public"
    PRIVATE = "private"
    SUBMISSION = "submission"


class EvidenceLevel(str, Enum):
    EXACT_DUPLICATE = "exact_duplicate"
    NEAR_DUPLICATE = "near_duplicate"
    PARTIAL_REUSE = "partial_reuse"
    POSSIBLE_MANIPULATION = "possible_manipulation"
    NEEDS_REVIEW = "needs_review"


def make_stable_id(prefix: str, *parts: str) -> str:
    payload = "\x1f".join(parts).encode("utf-8")
    digest = hashlib.sha256(payload).hexdigest()[:16]
    return f"{prefix}_{digest}"


@dataclass(frozen=True)
class ArticleRecord:
    article_id: str
    scope: CorpusScope
    title: str
    source_path: str
    source_sha256: str


@dataclass(frozen=True)
class FigureRecord:
    figure_id: str
    article_id: str
    page: int
    caption: str
    image_path: str
    bbox: tuple[float, float, float, float]


@dataclass(frozen=True)
class PanelRecord:
    panel_id: str
    figure_id: str
    label: str
    image_path: str
    bbox: tuple[float, float, float, float]


@dataclass(frozen=True)
class AnalysisFinding:
    finding_id: str
    level: EvidenceLevel
    target_panel_id: str
    message: str
    score: float
    algorithm: str
    source_panel_id: str | None = None

    def review_summary(self) -> str:
        return (
            f"{self.level.value}: {self.message} "
            f"via {self.algorithm} (score={self.score:.3f})"
        )


@dataclass(frozen=True)
class AICallRecord:
    call_id: str
    purpose: str
    model_id: str
    model_version: str
    prompt_template_id: str
    input_refs: tuple[str, ...]
    response_hash: str
    token_count: int
    privacy_filter_applied: bool
